src/input-capture.c

```diff
@@ -1154,74 +1202,74 @@ on_zones_changed_cb (XdpDbusImplInputCapture *impl,
 
-  xdp_dbus_input_capture_set_version (XDP_DBUS_INPUT_CAPTURE (input_capture), VERSION_1);
+  g_clear_object (&input_capture->impl);
 
-  supported_capabilities =
-    xdp_dbus_impl_input_capture_get_supported_capabilities (impl);
-  xdp_dbus_input_capture_set_supported_capabilities (XDP_DBUS_INPUT_CAPTURE (input_capture),
-                                                     supported_capabilities);
-
-  g_signal_connect (impl, "disabled", G_CALLBACK (on_disabled_cb), input_capture);
-  g_signal_connect (impl, "activated", G_CALLBACK (on_activated_cb), input_capture);
-  g_signal_connect (impl, "deactivated", G_CALLBACK (on_deactivated_cb), input_capture);
-  g_signal_connect (impl, "zones-changed", G_CALLBACK (on_zones_changed_cb), input_capture);
+  G_OBJECT_CLASS (input_capture_parent_class)->dispose (object);
 }
 
 static void
-input_capture_class_init (InputCaptureClass *klass)
+input_capture_init (InputCapture *input_capture)
 {
-  quark_request_session =
-    g_quark_from_static_string ("-xdp-request-capture-input-session");
 }
 
 static void
-input_capture_session_close (XdpSession *session)
+input_capture_class_init (InputCaptureClass *klass)
 {
-  InputCaptureSession *input_capture_session = INPUT_CAPTURE_SESSION (session);
-
-  input_capture_session->state = INPUT_CAPTURE_SESSION_STATE_CLOSED;
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
 
-  g_debug ("screen cast session owned by '%s' closed", session->sender);
-}
+  object_class->dispose = input_capture_dispose;
 
-static void
-input_capture_session_finalize (GObject *object)
-{
-  G_OBJECT_CLASS (input_capture_session_parent_class)->finalize (object);
-}
-
-static void
-input_capture_session_init (InputCaptureSession *input_capture_session)
-{
+  quark_request_session =
+    g_quark_from_static_string ("-xdp-request-capture-input-session");
 }
 
-static void
-input_capture_session_class_init (InputCaptureSessionClass *klass)
+static InputCapture *
+input_capture_new (XdpDbusImplInputCapture *impl)
 {
-  GObjectClass *object_class;
-  XdpSessionClass *session_class;
+  InputCapture *input_capture;
 
-  object_class = G_OBJECT_CLASS (klass);
-  object_class->finalize = input_capture_session_finalize;
-
-  session_class = (XdpSessionClass *)klass;
-  session_class->close = input_capture_session_close;
+  input_capture = g_object_new (input_capture_get_type (), NULL);
+  input_capture->impl = g_object_ref (impl);
+
+  g_dbus_proxy_set_default_timeout (G_DBUS_PROXY (input_capture->impl), G_MAXINT);
+
+  xdp_dbus_input_capture_set_version (XDP_DBUS_INPUT_CAPTURE (input_capture), 1);
+
+  g_object_bind_property (G_OBJECT (input_capture->impl), "supported-capabilities",
+                          G_OBJECT (input_capture), "supported-capabilities",
+                          G_BINDING_SYNC_CREATE);
+
+  g_signal_connect (input_capture->impl, "disabled",
+                    G_CALLBACK (on_disabled_cb),
+                    input_capture->impl);
```

Comment:
The old code passed in `input_capture` as userdata to the signal callbacks, this code now passes in `input_capture->impl`. But at quick glance it looks like the userdata is never actually used (it's passed into the callbacks, which pass it into `pass_signal`, which never uses it). So really you should just be passing in `NULL` as the userdata